sources/kms/custom-key-store.go

```diff
@@ -0,0 +1,153 @@
+			item.Health = sdp.Health_HEALTH_PENDING.Enum()
+		case types.ConnectionStateTypeDisconnected:
+			item.Health = sdp.Health_HEALTH_UNKNOWN.Enum()
+		case types.ConnectionStateTypeFailed:
+			item.Health = sdp.Health_HEALTH_ERROR.Enum()
+		case types.ConnectionStateTypeDisconnecting:
+			item.Health = sdp.Health_HEALTH_PENDING.Enum()
+		default:
+			return nil, &sdp.QueryError{
+				ErrorType:   sdp.QueryError_OTHER,
+				ErrorString: "unknown Connection State",
+			}
+		}
+
+		if customKeyStore.CloudHsmClusterId != nil {
+			// +overmind:link cloudhsmv2-cluster
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "cloudhsmv2-cluster",
+					Method: sdp.QueryMethod_GET,
+					Query:  *customKeyStore.CloudHsmClusterId,
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// Changing the CloudHSM cluster will affect the custom key store
+					In: true,
+					// Updating the custom key store will not affect the CloudHSM cluster
+					Out: false,
+				},
+			})
+		}
+
+		// TODO: Activate this after enabling get vpc by name
+		/*
+			if customKeyStore.XksProxyConfiguration != nil &&
+				customKeyStore.XksProxyConfiguration.VpcEndpointServiceName != nil {
+				// +overmind:link ec2-vpc
+				item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+					Query: &sdp.Query{
+						Type:   "ec2-vpc",
+						Method: sdp.QueryMethod_GET,
+						Query:  *customKeyStore.XksProxyConfiguration.VpcEndpointServiceName,
+						Scope:  scope,
+					},
+					BlastPropagation: &sdp.BlastPropagation{
+						// Changing the VPC will affect the custom key store
+						In: true,
+						// Updating the custom key store will not affect the VPC
+						Out: false,
+					},
+				})
+			}
+		*/
+
+		items = append(items, &item)
+	}
+
+	return items, nil
+}
+
+//go:generate docgen ../../docs-data
+// +overmind:type kms-custom-key-store
+// +overmind:descriptiveType Custom Key Store
+// +overmind:get Get a custom key store by its ID or name
+// +overmind:list List all custom key stores
+// +overmind:search Search custom key store by ARN
+// +overmind:group AWS
+// +overmind:terraform:queryMap aws_kms_custom_key_store.id
+
+func NewCustomKeyStoreSource(client *kms.Client, accountID string, region string) *sources.DescribeOnlySource[*kms.DescribeCustomKeyStoresInput, *kms.DescribeCustomKeyStoresOutput, *kms.Client, *kms.Options] {
+	return &sources.DescribeOnlySource[*kms.DescribeCustomKeyStoresInput, *kms.DescribeCustomKeyStoresOutput, *kms.Client, *kms.Options]{
+		Region:    region,
+		Client:    client,
+		AccountID: accountID,
+		ItemType:  "kms-custom-key-store",
+		DescribeFunc: func(ctx context.Context, client *kms.Client, input *kms.DescribeCustomKeyStoresInput) (*kms.DescribeCustomKeyStoresOutput, error) {
+			return client.DescribeCustomKeyStores(ctx, input)
+		},
+		InputMapperGet: func(_, query string) (*kms.DescribeCustomKeyStoresInput, error) {
```

Comment:
The `GET` method should always take the value of the `uniqueAttributeValue` as it's param, and nothing else. So in this case the UAV is `customKeyStoreId`, so GET should accept just a `customKeyStoreId`. Then if it makes sense to search by name (like if it's reference by name in other resources) I'd recommend implementing that as a custom search. You can always have it searchable by ARN too by just trying to parse the query as an ARN and changing the logic based on that.

Reply:
Currently search by name is not required for linking purpose. I'll change it to it only. I think search by ARN is sufficient for how.  

Thanks for this.

Reply:
https://github.com/overmindtech/aws-source/pull/571